File: fleet/runtime/comb.ts
// The Comb memory seam. Comb is the hive's durable, provenance-gated knowledge
// graph (Go core/kmstore, served by services/comb-mcp and driven by the
// `agix-core km` CLI). A TS agent grows and reads governed knowledge through this
// interface — reads are attested-only by default (the provenance-first posture),
// writes carry author + a DISTINCT verifier + a trust score. The attestation gate
// and anti-poison shield live in Go; this client never reimplements them.
//
//   - CliComb  — production: shells `agix-core km …` against the shared store
//                (~/.agix/km.db, the same store the Comb MCP serves).
//   - MemComb  — an in-memory store for `bun test` (attestation modeled by the
//                same actor≠verifier + trust-floor rule the Go gate enforces).
//
// A third impl, McpComb (JSON-RPC to `comb-mcp -stdio`), is the networked fleet
// seam; it satisfies this same interface. Left as a documented extension.
//
// Copyright 2026 Agix AI LLC. Apache-2.0.

export interface Provenance {
  author?: string;
  /** A DISTINCT, registered actor that vouches (empty → un-attested). */
  verifier?: string;
  /** Verifier confidence 0..1; must clear the store's trust floor to attest. */
  trust?: number;
}

export interface CombNote extends Provenance {
  content: string;
  id?: string;
  branch?: string;
  ratified?: boolean;
}

export interface CombPutResult {
  id: string;
  added: boolean;
  attested: boolean;
  quarantined: boolean;
  reason?: string;
}

export interface CombLeaf {
  id: string;
  content: string;
  branch?: string;
  attested: boolean;
}

export interface CombStats {
  leaves: number;
  attested: number;
  ratified: number;
  edges: number;
  quarantined: number;
  trustFloor: number;
}

export interface Comb {
  put(note: CombNote): Promise<CombPutResult>;
  link(src: string, type: string, dst: string, prov?: Provenance): Promise<void>;
  retrieve(query: string, k?: number, attestedOnly?: boolean): Promise<CombLeaf[]>;
  traverse(seed: string, type: string, hops?: number): Promise<CombLeaf[]>;
  stats(): Promise<CombStats>;
}

/** CliComb drives the Go provenance-gated store via `agix-core km`. It parses the
 *  CLI's deterministic, documented output. The attestation roster is seeded out of
 *  band in Go (AGIX_KM_VERIFIERS), so a bare --verifier cannot forge attestation —
 *  this client cannot bypass that gate. */
export class CliComb implements Comb {
  private readonly bin: string;
  private readonly db?: string;

  constructor(opts: { bin?: string; db?: string } = {}) {
    this.bin = opts.bin ?? Bun.env.AGIX_CORE_BIN ?? "agix-core";
    this.db = opts.db ?? Bun.env.COMB_MCP_STORE ?? undefined;
  }

  private async km(sub: string, args: string[]): Promise<string> {
    const argv = ["km", sub, ...args];
    if (this.db) argv.push("--db", this.db);
    const proc = Bun.spawn([this.bin, ...argv], { stdout: "pipe", stderr: "pipe" });
    const [out, err, code] = await Promise.all([
      new Response(proc.stdout).text(),
      new Response(proc.stderr).text(),
      proc.exited,
    ]);
    // km returns 3 for a shielded (quarantined) write — not a hard failure.
    if (code !== 0 && code !== 3) throw new Error(`comb: km ${sub} failed (exit ${code}): ${err.trim()}`);
    return out;
  }

  async put(note: CombNote): Promise<CombPutResult> {
    const args = ["--content", note.content];
    if (note.id) args.push("--id", note.id);
    if (note.branch) args.push("--branch", note.branch);
    if (note.author) args.push("--author", note.author);
    if (note.verifier) args.push("--verifier", note.verifier);
    if (typeof note.trust === "number") args.push("--trust", String(note.trust));
    if (note.ratified) args.push("--ratified");
    const out = await this.km("put", args);
    const m = out.match(/id=(\S+)\s+added=(\w+)\s+attested=(\w+)\s+quarantined=(\w+)/);
    if (!m) throw new Error(`comb: unparseable km put output: ${out}`);
    return { id: m[1], added: m[2] === "true", attested: m[3] === "true", quarantined: m[4] === "true" };
  }

  async link(src: string, type: string, dst: string, prov: Provenance = {}): Promise<void> {
    const args = ["--src", src, "--type", type, "--dst", dst];
    if (prov.author) args.push("--author", prov.author);
    if (prov.verifier) args.push("--verifier", prov.verifier);
    if (typeof prov.trust === "number") args.push("--trust", String(prov.trust));
    await this.km("link", args);
  }

  async retrieve(query: string, k = 5, attestedOnly = true): Promise<CombLeaf[]> {
    const args = ["--query", query, "--k", String(k)];
    if (attestedOnly) args.push("--attested-only");
    const out = await this.km("retrieve", args);
    return parseLeafLines(out);
  }

  async traverse(seed: string, type: string, hops = 1): Promise<CombLeaf[]> {
    const out = await this.km("traverse", ["--seed", seed, "--type", type, "--hops", String(hops), "--attested-only"]);
    return parseLeafLines(out);
  }

  async stats(): Promise<CombStats> {
    const out = await this.km("stats", []);
    const m = out.match(
      /leaves=(\d+)\s+attested=(\d+)\s+ratified=(\d+)\s+tombstoned=\d+\s+edges=(\d+)\s+quarantined=(\d+)\s+trust_floor=([\d.]+)/,
    );
    if (!m) throw new Error(`comb: unparseable km stats output: ${out}`);
    return {
      leaves: +m[1],
      attested: +m[2],
      ratified: +m[3],
      edges: +m[4],
      quarantined: +m[5],
      trustFloor: +m[6],
    };
  }
}

// km retrieve/traverse print numbered lines like:
//   1. <id>  branch=<b> attested=<bool>  "<content>"
function parseLeafLines(out: string): CombLeaf[] {
  const leaves: CombLeaf[] = [];
  for (const line of out.split("\n")) {
    const m = line.match(/^\s*\d+\.\s+(\S+)\s+(?:branch=(\S+)\s+)?attested=(\w+)\s+"(.*)"\s*$/);
    if (m) leaves.push({ id: m[1], branch: m[2] === "-" ? undefined : m[2], attested: m[3] === "true", content: m[4] });
  }
  return leaves;
}

/** MemComb is an in-memory Comb for tests. It models the SAME governance the Go
 *  gate enforces: a leaf is attested only when a verifier DISTINCT from the author,
 *  present on the roster, vouches with trust ≥ the floor; retrieve(attestedOnly)
 *  refuses un-attested knowledge. Retrieval is a substring match (a stand-in for
 *  the cosine-similar embedding search), which is enough to prove agent wiring. */
export class MemComb implements Comb {
  private leaves = new Map<string, { note: CombNote; attested: boolean }>();
  private edges: { src: string; type: string; dst: string; attested: boolean }[] = [];
  private roster: Set<string>;
  private floor: number;

  constructor(opts: { roster?: string[]; trustFloor?: number } = {}) {
    this.roster = new Set(opts.roster ?? []);
    this.floor = opts.trustFloor ?? 0.35;
  }

  registerVerifier(...actors: string[]): void {
    for (const a of actors) this.roster.add(a);
  }

  private attests(note: CombNote): boolean {
    return (
      !!note.verifier &&
      note.verifier !== note.author &&
      this.roster.has(note.verifier) &&
      (note.trust ?? 0) >= this.floor
    );
  }

  async put(note: CombNote): Promise<CombPutResult> {
    const id = note.id ?? `leaf-${hash(note.content)}`;
    const attested = this.attests(note);
    const prior = this.leaves.get(id);
    // Idempotent refresh only ever RAISES attestation (mirrors kmstore).
    this.leaves.set(id, { note: { ...note, id }, attested: attested || (prior?.attested ?? false) });
    return { id, added: !prior, attested: this.leaves.get(id)!.attested, quarantined: false };
  }

  async link(src: string, type: string, dst: string, prov: Provenance = {}): Promise<void> {
    this.edges.push({
      src,
      type,
      dst,
      attested: !!prov.verifier && prov.verifier !== prov.author && this.roster.has(prov.verifier) && (prov.trust ?? 0) >= this.floor,
    });
  }

  async retrieve(query: string, k = 5, attestedOnly = true): Promise<CombLeaf[]> {
    const q = query.toLowerCase();
    const hits: CombLeaf[] = [];
    for (const [id, { note, attested }] of this.leaves) {
      if (attestedOnly && !attested) continue;
      if (note.content.toLowerCase().includes(q) || q.includes(id.toLowerCase())) {
        hits.push({ id, content: note.content, branch: note.branch, attested });
      }
    }
    return hits.slice(0, k);
  }

  async traverse(seed: string, type: string, hops = 1): Promise<CombLeaf[]> {
    const reached: CombLeaf[] = [];
    let frontier = [seed];
    for (let h = 0; h < hops; h++) {
      const next: string[] = [];
      for (const e of this.edges) {
        if (e.attested && e.type === type && frontier.includes(e.src)) {
          const leaf = this.leaves.get(e.dst);
          if (leaf?.attested) {
            reached.push({ id: e.dst, content: leaf.note.content, branch: leaf.note.branch, attested: true });
            next.push(e.dst);
          }
        }
      }
      frontier = next;
    }
    return reached;
  }

  async stats(): Promise<CombStats> {
    let attested = 0;
    let ratified = 0;
    for (const { note, attested: a } of this.leaves.values()) {
      if (a) attested++;
      if (note.ratified) ratified++;
    }
    return {
      leaves: this.leaves.size,
      attested,
      ratified,
      edges: this.edges.length,
      quarantined: 0,
      trustFloor: this.floor,
    };
  }
}

function hash(s: string): string {
  let h = 5381;
  for (let i = 0; i < s.length; i++) h = ((h << 5) + h + s.charCodeAt(i)) >>> 0;
  return h.toString(36);
}

File: core/comb/comb.go
// Package comb is the developer-facing Go SDK over the Comb — the hive's durable
// provenance-gated knowledge store (kmstore). It is the ergonomic client the MCP
// server (services/comb-mcp) transports and the swarm's CombRetriever consumes
// under the hood, but with a provenance-FIRST surface: reads are attested-only by
// default, writes carry their author/verifier/trust, and two value-adds the raw
// store does not offer — a queryable LINEAGE / bee-trace walk (author → verifier
// → … → human root) and a self-contained HTML-Comb artifact emitter.
//
// This package only READS from kmstore / apiary / caste / ledger; it never
// reimplements the attestation gate or the anti-poisoning shield — those stay in
// kmstore. It is a thin, stdlib-plus-core-leaf facade: the value is ergonomics
// plus lineage.go and html.go, not a 1:1 re-wrap.
//
// Copyright 2026 Agix AI LLC. Apache-2.0.
package comb

import (
	"errors"

	"github.com/agix-ai/agix/core/kmstore"
	"github.com/agix-ai/agix/core/ledger"
)

// DefaultDim is the embedding dimension the SDK hashes text at when a Note does
// not carry its own vector. It matches core/cmd/agix-core/km.go's embedDim (64)
// and the swarm CombRetriever's dim, so a note written through this SDK is
// byte-identically retrievable by the `km` CLI and by worker foraging — a drift
// would silently return nothing on retrieval (see kmstore.HashEmbed's contract).
const DefaultDim = 64

// errNoStore guards the nil-store paths so a misconstructed Comb fails loudly
// rather than panicking on a nil dereference.
var errNoStore = errors.New("comb: nil store (build with comb.New(store))")

// Comb is the ergonomic, provenance-first client facade over a *kmstore.KMStore.
// It is safe for concurrent use to the same degree kmstore is (WAL-mode reads,
// single-writer). Build one with New and, optionally, an audit ledger to enrich
// lineage traces with the trail of record.
type Comb struct {
	store *kmstore.KMStore
	led   *ledger.Ledger // optional: corroborates lineage hops with recorded frames
	dim   int            // embedding dimension for text-query / text-write ergonomics
}

// Option configures a Comb at construction.
type Option func(*Comb)

// WithDim overrides the embedding dimension (default DefaultDim). It MUST equal
// the dim the store's existing leaves were written at, or text retrieval misses.
func WithDim(dim int) Option {
	return func(c *Comb) {
		if dim > 0 {
			c.dim = dim
		}
	}
}

// WithLedger attaches the append-only audit ledger so TraceLeaf / TraceActor can
// corroborate each provenance hop with concrete recorded frames (agent_start,
// model_call, ratify …) and, heuristically, resolve the human principal from a
// ratify frame. Lineage still reconstructs without it — just less grounded.
func WithLedger(l *ledger.Ledger) Option {
	return func(c *Comb) { c.led = l }
}

// New builds a Comb over an already-open store. The store owns its lifecycle
// (Open/Close); this facade never closes it.
func New(store *kmstore.KMStore, opts ...Option) *Comb {
	c := &Comb{store: store, dim: DefaultDim}
	for _, o := range opts {
		o(c)
	}
	if c.dim <= 0 {
		c.dim = DefaultDim
	}
	return c
}

// Dim reports the embedding dimension this client hashes text at.
func (c *Comb) Dim() int { return c.dim }

// RegisterVerifier adds principals to the store's attestation roster — the
// out-of-band allowlist of identities authorized to vouch for a write (the fix
// for forgeable attestation; see kmstore/provenance.go). A verifier named on a
// Put/Attest is only honored if it was registered here first.
func (c *Comb) RegisterVerifier(actors ...string) {
	if c == nil || c.store == nil {
		return
	}
	c.store.RegisterVerifier(actors...)
}

// TrustFloor returns the store's attestation trust floor in force.
func (c *Comb) TrustFloor() float64 {
	if c == nil || c.store == nil {
		return 0
	}
	return c.store.TrustFloor()
}

// Note is the ergonomic input for writing one leaf into the Comb. Content is the
// text payload; Author wrote it; Verifier (a distinct, REGISTERED actor) attests
// it; Trust is the verifier's confidence in 0..1. Leave Embedding nil to hash
// Content at the client's dim; supply it to carry a real model's vector.
type Note struct {
	ID            string    // optional; empty → content-derived id (dedupes identical knowledge)
	Content       string    // the leaf's text payload
	Branch        string    // optional TOGAF branch tag
	Author        string    // actor ref that asserted the note, e.g. "agix/worker/forager-1"
	Verifier      string    // distinct, registered actor that vouches (empty → un-attested)
	Trust         float64   // verifier confidence, 0..1 (must clear TrustFloor to attest)
	Ratified      bool      // operator-ratified (trunk-merge) bit
	PendingCosign bool      // un-attested leaf awaiting a human co-sign (judgment-only verdict)
	Embedding     []float32 // optional; nil → kmstore.HashEmbed(Content, dim)
}

func (n Note) leaf(dim int) kmstore.Leaf {
	emb := n.Embedding
	if emb == nil {
		emb = kmstore.HashEmbed(n.Content, dim)
	}
	return kmstore.Leaf{
		ID:            n.ID,
		Content:       n.Content,
		Branch:        n.Branch,
		Author:        n.Author,
		Verifier:      n.Verifier,
		TrustScore:    n.Trust,
		Ratified:      n.Ratified,
		PendingCosign: n.PendingCosign,
		Embedding:     emb,
	}
}

// Put writes a note through kmstore's provenance gate + anti-poisoning shield and
// reports the outcome (added / attested / quarantined + a reason). Attestation is
// decided inline by the actor≠verifier rule — a note becomes attested iff a
// distinct, registered Verifier vouches with Trust ≥ TrustFloor. A note that
// fails the gate is stored UN-attested (retrievable only via RetrieveAll), never
// silently rejected — unless it contradicts an existing attested leaf, in which
// case the shield quarantines it (PutResult.Quarantined).
func (c *Comb) Put(n Note) (kmstore.PutResult, error) {
	if c == nil || c.store == nil {
		return kmstore.PutResult{}, errNoStore
	}
	return c.store.Put(n.leaf(c.dim))
}

// Attest raises an EXISTING leaf to attested by having a distinct, registered
// verifier vouch for it. kmstore keys attestation on content identity and its
// idempotent-refresh path only ever RAISES attestation (never downgrades), so
// this is a re-Put of the same Content carrying the vouching Verifier. Because
// the store exposes no fetch-by-id, the caller must supply the leaf's Content
// (and its original Author, so the actor≠verifier gate still holds). The returned
// PutResult.Attested / .Reason report whether the vouch landed.
func (c *Comb) Attest(n Note) (kmstore.PutResult, error) {
	if c == nil || c.store == nil {
		return kmstore.PutResult{}, errNoStore
	}
	if n.Verifier == "" {
		return kmstore.PutResult{}, errors.New("comb: Attest needs a Verifier (a second actor must vouch)")
	}
	return c.store.Put(n.leaf(c.dim))
}

// Cosign promotes an EXISTING pending-cosign leaf to attested by having a
// distinct, registered human verifier vouch for it — the manual half of the
// attestation policy (a judgment-only governed verdict was held out of the
// corpus; a human now certifies it). It follows the same actor≠verifier gate as
// a write, so the co-signer must be registered, distinct from the leaf's author,
// and clear the trust floor. The returned PutResult.Attested reports whether the
// co-sign landed.
func (c *Comb) Cosign(id, verifier string, trust float64) (kmstore.PutResult, error) {
	if c == nil || c.store == nil {
		return kmstore.PutResult{}, errNoStore
	}
	if verifier == "" {
		return kmstore.PutResult{}, errors.New("comb: Cosign needs a verifier (a human must vouch)")
	}
	return c.store.Cosign(id, verifier, trust)
}

// Link asserts a typed edge src -[edgeType]-> dst carrying its own provenance.
// The edge is attested by the same actor≠verifier rule as leaves; an un-attested
// edge is skipped by governed (attested-only) Traverse.
func (c *Comb) Link(src, edgeType, dst string, prov kmstore.Provenance) error {
	if c == nil || c.store == nil {
		return errNoStore
	}
	return c.store.Link(src, edgeType, dst, prov)
}

// Retrieve is the GOVERNED read: it embeds the query text and returns the k
// most cosine-similar ATTESTED leaves (un-attested / quarantined / tombstoned
// knowledge is structurally unreachable). This is the provenance-first default —
// the same posture the swarm's CombRetriever forages under.
func (c *Comb) Retrieve(query string, k int) ([]kmstore.Leaf, error) {
	return c.retrieve(query, k, true)
}

// RetrieveAll is the UN-governed read: it returns the k most similar leaves
// regardless of attestation (still excluding tombstoned). Use it for audit /
// rendering — e.g. to show attested vs un-attested state in an HTML artifact —
// never for foraging.
func (c *Comb) RetrieveAll(query string, k int) ([]kmstore.Leaf, error) {
	return c.retrieve(query, k, false)
}

func (c *Comb) retrieve(query string, k int, attestedOnly bool) ([]kmstore.Leaf, error) {
	if c == nil || c.store == nil {
		return nil, errNoStore
	}
	vec := kmstore.HashEmbed(query, c.dim)
	return c.store.Retrieve(vec, k, kmstore.RetrieveOpts{AttestedOnly: attestedOnly})
}

// CertifiedLeaves enumerates EVERY attested leaf on branch (all of it, not a
// cosine top-k) — the corpus source for distillation. Only verifier-vouched
// (actor≠verifier, trust ≥ floor) knowledge is returned, which is exactly the
// clean training signal the flywheel distills: plausible-but-wrong edits are never
// attested, so they never reach the weights. Pass "" for branch to enumerate all
// attested leaves; refactoring records live on the "software" branch.
func (c *Comb) CertifiedLeaves(branch string) ([]kmstore.Leaf, error) {
	if c == nil || c.store == nil {
		return nil, errNoStore
	}
	return c.store.Leaves(kmstore.LeafFilter{Branch: branch, AttestedOnly: true})
}

// Traverse follows attested edges of edgeType from seedID for hops steps and
// returns the reached attested leaves in breadth-first discovery order (seed
// excluded). It is always governed — un-attested edges and leaves are refused.
func (c *Comb) Traverse(seedID, edgeType string, hops int) ([]kmstore.Leaf, error) {
	if c == nil || c.store == nil {
		return nil, errNoStore
	}
	return c.store.Traverse(seedID, edgeType, hops, kmstore.TraverseOpts{AttestedOnly: true})
}

// Stats returns a snapshot of the store's contents and governing floor.
func (c *Comb) Stats() (kmstore.Stats, error) {
	if c == nil || c.store == nil {
		return kmstore.Stats{}, errNoStore
	}
	return c.store.Stats()
}
